### PR TITLE
Support Swift bridging for RACSignalError using NS_ERROR_ENUM

### PR DESCRIPTION
**Pull request by sharplet**
Redefines `RACSignalError` using `NS_ERROR_ENUM`, allowing it to bridge to Swift as a new `RACSignalError` struct containing static instances of `RACSignalError.Code`, and allowing it to be used in error pattern matching contexts.

### CODE REVIEW

**Comment by sharplet** [comment]
There’s a potential breaking change here: the symbols `RACSignalErrorNoMatchingCase` and `RACSignalErrorTimedOut` will no longer be available to Swift. I pushed commit 2334221, which adds them back in with a deprecated+renamed warning. Any thoughts on whether this is necessary? The downside is that it adds Swift code to a previously ObjC-only framework, which could cause build errors.

**Comment by sharplet** [comment]
Discoved `RACCommandErrorDomain` and `RACSelectorSignalErrorDomain` and added bridging for those as well.

**Comment by sharplet** [comment]
The `pod lib lint` failure appears to be a CocoaPods bug (https://github.com/CocoaPods/CocoaPods/issues/8089):

```
 -> ReactiveObjC (3.1.0)
    - NOTE  | xcodebuild:  note: Using new build system
    - NOTE  | xcodebuild:  note: Planning build
    - NOTE  | xcodebuild:  note: Constructing build description
    - NOTE  | xcodebuild:  warning: Capabilities for App may not function correctly because its entitlements use a placeholder team ID. To resolve this, select a development team in the build settings editor. (in target 'App')
    - NOTE  | xcodebuild:  warning: Skipping code signing because the target does not have an Info.plist file. (in target 'App')
    - ERROR | [OSX] xcodebuild: Returned an unsuccessful exit code. You can use `--verbose` for more information.
```

Updating to CocoaPods 1.6.0.beta.2 locally resolved the error for me.

**Comment by sharplet** [comment]
For some reason, switching to `Apple TV 4K` instead of `Apple TV 1080p` worked, even though `Apple TV 1080p` clearly seemed to be in the list of available destinations.

The CocoaPods thing is still there, but I’m comfortable ignoring it until 1.6.0 is released.

File: ReactiveObjC.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -2579,13 +2600,19 @@
 					"DTRACE_PROBES_DISABLED=1",
 				);
 				INFOPLIST_FILE = ReactiveObjC/Info.plist;
+				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/../Frameworks @loader_path/Frameworks";
```
Comment (mdiep) [inline ReactiveObjC.xcodeproj/project.pbxproj]: This shouldn't be needed, I don't think. It should already be specified in the `xcconfig`?

File: ReactiveObjC/RACBlockTrampoline.m [hunk 2]
```diff
@@ -28,7 +28,7 @@ - (instancetype)initWithBlock:(id)block {
 + (id)invokeBlock:(id)block withArguments:(RACTuple *)arguments {
 	NSCParameterAssert(block != NULL);
 
-	RACBlockTrampoline *trampoline = [[self alloc] initWithBlock:block];
+	RACBlockTrampoline *trampoline = [(RACBlockTrampoline *)[self alloc] initWithBlock:block];
```
Comment (mdiep) [inline ReactiveObjC/RACBlockTrampoline.m]: Why was this needed?

File: ReactiveObjC/Deprecations+Removals.swift [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+public let RACCommandErrorNotEnabled = RACCommandError.notEnabled.rawValue
+
+@available(*, deprecated, renamed: "RACSignalError.methodSwizzlingRace")
+public let RACSelectorSignalErrorMethodSwizzlingRace = RACSelectorSignalError.methodSwizzlingRace
+
+@available(*, deprecated, renamed: "RACSignalError.noMatchingCase")
+public let RACSignalErrorNoMatchingCase = RACSignalError.noMatchingCase.rawValue
+
+@available(*, deprecated, renamed: "RACSignalError.timedOut")
+public let RACSignalErrorTimedOut = RACSignalError.timedOut.rawValue
```
Comment (mdiep) [inline ReactiveObjC/Deprecations+Removals.swift]: This seems great. 👏

File: ReactiveObjC.xcodeproj/project.pbxproj (hunk 1 above)
Comment (sharplet) [reply ReactiveObjC.xcodeproj/project.pbxproj]: Right, this was probably an automatic Xcode change when adding the first Swift code to the target.

File: ReactiveObjC/RACBlockTrampoline.m (hunk 2 above)
Comment (sharplet) [reply ReactiveObjC/RACBlockTrampoline.m]: While debugging the [`pod lib lint` failures on CI](https://travis-ci.org/ReactiveCocoa/ReactiveObjC/jobs/472432768) I noticed warnings like this:

```
- WARN  | [iOS] xcodebuild:  /Users/travis/build/ReactiveCocoa/ReactiveObjC/ReactiveObjC/RACBlockTrampoline.m:31:49: warning: 'initWithBlock:' is only available on iOS 10.0 or newer [-Wunguarded-availability]
```

This is a false positive, because it’s confusing our `-[RACBlockTrampoline initWithBlock:]` selector for something like `-[NSThread initWithBlock:]`, which has only been available since iOS 10. Adding the explicit cast shouldn’t be necessary (this seems like a compiler/analyser bug), but it does remove the warning.

File: ReactiveObjC.xcodeproj/project.pbxproj (hunk 1 above)
Comment (sharplet) [reply ReactiveObjC.xcodeproj/project.pbxproj]: I removed all the redundant `LD_RUNPATH_SEARCH_PATHS` settings in commit 31e9f4aca5705c0f1f05df1f3340a4a869b2dbcb.

File: ReactiveObjC.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -2895,9 +2980,11 @@
 			baseConfigurationReference = D047263219E49FE8006002AA /* iOS-Application.xcconfig */;
 			buildSettings = {
 				ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES = YES;
+				CLANG_ENABLE_MODULES = YES;
 				INFOPLIST_FILE = ReactiveObjCTests/Info.plist;
-				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks @loader_path/Frameworks";
 				PRODUCT_NAME = "$(PROJECT_NAME)Tests";
+				SWIFT_OPTIMIZATION_LEVEL = "-Onone";
+				SWIFT_VERSION = 4.2;
```
Comment (sharplet) [inline ReactiveObjC.xcodeproj/project.pbxproj]: We could potentially lower this version to 4.0 or even 3.0 to keep support for compiling with older Xcode versions.

Comment (mdiep) [reply ReactiveObjC.xcodeproj/project.pbxproj]: I'm not sure that's important.

**Comment by mdiep** [comment]
This looks good once the `pod lib lint` is fixed.

**Comment by mdiep** [comment]
I'm fine with that. 👍🏻 Merge at your leisure!

**Comment by sharplet** [comment]
Oh weird, looks like I don’t have write access for ReactiveObjC?

**Comment by sharplet** [comment]
@mdiep Thanks!
